Popup positioning rework

kekekeks opened the pull request:
Added and implemented IPopupPositioner as described in https://github.com/AvaloniaUI/Avalonia/issues/2405

Aside from being wayland compatible and being decoupled from Popup/PopupImpl logic, the new positioning infrastructure is actually aware of screen bounds and is also aware of placement rectangles when trying to fit the popup into the screen, so the placement target won't be obscured by the popup.

![output](https://user-images.githubusercontent.com/1067584/61994824-5800d080-b088-11e9-9d75-0f0f7da58dfc.gif)



Breaking changes:
- Popup now needs either a PlacementTarget or a logical parent attached to the visual tree.
- PopupRoot now always needs a toplevel parent.

kekekeks committed:
Make toplevels responsible of creating popups

kekekeks committed:
PixelVector and stuff

kekekeks committed:
IPopupPositioner + managed implementation

kekekeks committed:
Wired up the popup positioner

Tests are failing because they are trying create popups out of a thin air

kekekeks committed:
Somewhat fixed tests

kekekeks committed:
Completely re-create PopupRoot on reopening Popup

kekekeks committed:
Dispose GL drawing session if we were unable to create a drawing context

kekekeks committed:
Fixed tests

kekekeks committed:
Don't scale the  screen in ManagedPopupPositionerPopupImplHelper

kekekeks committed:
Pass the non-translated size to the popup

kekekeks committed:
Removed old debug code

kekekeks committed:
Removed android popup implementation since it was broken anyway

kekekeks committed:
Dead code

kekekeks committed:
Build

grokys commented on src/Avalonia.Controls/Primitives/Popup.cs:
```diff
@@ -218,57 +226,71 @@ public bool Topmost
+            _popupRoot = new PopupRoot(_topLevel, DependencyResolver)
             {
-                _topLevel = PlacementTarget.GetSelfAndLogicalAncestors().First(x => x is TopLevel) as TopLevel;
+                [~WidthProperty] = this[~WidthProperty],
```
We probably want to rethink binding the popup root's sizing parameters to `Popup`: see https://github.com/AvaloniaUI/Avalonia/issues/2373

Doesn't need to be done in this PR, but thought I'd make you aware of that issue.

kekekeks replied:
We probably need to make `Popup` to not be a part of the visual tree. Then we could derive from StyledElement directly and define new Width/Height properties

grokys commented on src/Avalonia.Controls/Primitives/Popup.cs:
```diff
@@ -492,5 +470,54 @@ public void Dispose()
             }
         }
+
+        // For some reason when PopupRoot.Content is bound directly to Child weird stuff happens 
```
Hmm, what weird stuff?

kekekeks replied:
Yes, for some reason the content gets stripped of all of its children.

kekekeks replied:
I've initially added `PopupContentHost` to debug the visual children disappearance, but it has suddenly stopped to be a problem, so I've left it as is.

kekekeks replied:
I think I will need a class like that for overlay popups, so I think we should keep it.

kekekeks committed:
IPopupImpl is now optional advanced feature

kekekeks committed:
watwut

kekekeks commented on tests/Avalonia.Controls.UnitTests/Primitives/PopupTests.cs:
```diff
@@ -233,46 +223,28 @@ public void Popup_Close_Should_Raise_Single_Closed_Event()
         {
+            if(UsePopupHost)
+                // For some reason with overlay popups templates don't get applied in test mode but
+                // everything works perfectly fine at runtime. I leave this one to you @grokys
```
Actually tooltips and DatePicker are broken, so it's not "perfectly fine"

kekekeks committed:
Defensive check for Menu==null

grokys commented on src/Avalonia.Controls/Primitives/IPopupHost.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public interface IPopupHost : IDisposable
+    {
+        void SetContent(IControl control);
```
`Content` usually refers to data that can be anything, e.g. `object Content`. If the content is a control, the property is usually named `Child`.

grokys commented on src/Avalonia.Controls/Primitives/IPopupHost.cs:
```diff
@@ -0,0 +1,22 @@
+    public interface IPopupHost : IDisposable
+    {
+        void SetContent(IControl control);
+        IVisual VisualRoot { get; }
```
Why is this property needed? It returns `this` in both implementations of this class.

kekekeks replied:
It returns `null` for `PopupHost`. It's needed for devtools, I believe.

grokys commented on src/Avalonia.Controls/Primitives/OverlayLayer.cs:
```diff
@@ -0,0 +1,146 @@
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            
+            var infinite = new Size(double.PositiveInfinity, double.PositiveInfinity);
```
There's already `Size.Infinite`.

kekekeks committed:
OverlayLayer is now a glorified Canvas

kekekeks committed:
Added custom hit-testing for ImmediateRenderer to OverlayLayer

kekekeks committed:
Remove misleading braces

grokys committed:
Make `PopupHost` a templated control.

grokys committed:
Reset inheritance parent in ContentPresenter.

Previously a dangling reference to `ContentPresenter`  was left in place in the ex-child control's `InheritanceParent`. When the child is removed from the `ContentPresenter`, reset the value to the control's logical parent.

grokys committed:
Remove PopupContentHost.

It's no longer needed now the bug fix in 669c6511d6158c6f7440ea4927f36971134251d1 is in place.

grokys committed:
Move setting popup TemplatedParent...

...out of `PopupRoot` and into `Popup`. Fixes styling problems in `Popups` hosted in control templates.

grokys committed:
Fix IsInsidePopup logic.

grokys commented on src/Avalonia.Controls/Primitives/VisualLayerManager.cs:
```diff
@@ -0,0 +1,97 @@
+        private const int OverlayZIndex = int.MaxValue - 99;
+
+        private bool _isAttachedToLogicalTree;
+        private IStyleHost _styleHost;
```
Nit: I think this should be called `_styleRoot`.

grokys commented on src/Avalonia.Controls/Primitives/VisualLayerManager.cs:
```diff
@@ -0,0 +1,97 @@
+
+        private bool _isAttachedToLogicalTree;
+        private IStyleHost _styleHost;
+        public bool IsPopup { get; set; }
```
Nit: Can we move this to be with other properties? It gets mistaken as a private field here.

grokys commented on src/Avalonia.Controls/Primitives/VisualLayerManager.cs:
```diff
@@ -0,0 +1,97 @@
+        private IStyleHost _styleHost;
+        public bool IsPopup { get; set; }
+        
+        List<Control> _layers = new List<Control>();
```
Nit: missing `private readonly`.

grokys commented on src/Avalonia.Controls/Primitives/PopupPositioning/ManagedPopupPositioner.cs:
```diff
@@ -0,0 +1,175 @@
+        }
+
+        
+        void Update(Size translatedSize, Size originalSize,
```
Nit: missing `private`.

grokys commented on src/Avalonia.Themes.Default/ComboBox.xaml:
```diff
@@ -39,7 +39,7 @@
               <Border BorderBrush="{DynamicResource ThemeBorderMidBrush}"
                       BorderThickness="1">
-                  <AdornerDecorator Margin="-1 -1 0 0">
+                  <VisualLayerManager IsPopup="True" Margin="-1 -1 0 0">
```
Now that `PopupHost` is a templated control too, we can probably move this into the template?

grokys commented on src/Avalonia.Visuals/Media/PixelPoint.cs:
```diff
@@ -106,6 +159,8 @@ public override int GetHashCode()
         }
+        
+        
 
```
Nit: whitespace changes.

grokys commented on src/Avalonia.Visuals/Media/PixelRect.cs:
```diff
@@ -261,6 +261,17 @@ public bool Intersects(PixelRect rect)
+            return new PixelRect(Position + offset, Size);
+        }
+
 
```
Nit: whitespace.

grokys commented on src/Avalonia.Visuals/Media/PixelVector.cs:
```diff
@@ -0,0 +1,205 @@
+        /// <returns></returns>
+        public bool Equals(PixelVector other)
+        {
+            // ReSharper disable CompareOfFloatsByEqualityOperator
```
I think we can remove these resharper pragmas as we're not comparing floats here.

kekekeks committed:
Review comments
